core/src/main/java/com/oath/oak/OakRKeyBufferImpl.java

```diff
@@ -6,86 +6,149 @@
 
 package com.oath.oak;
 
+import com.oath.oak.NativeAllocator.OakNativeMemoryAllocator;
+import sun.nio.ch.DirectBuffer;
+
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
 import java.util.function.Function;
 
-import static com.oath.oak.Chunk.BLOCK_ID_LENGTH_ARRAY_INDEX;
-import static com.oath.oak.Chunk.KEY_BLOCK_SHIFT;
-import static com.oath.oak.Chunk.KEY_LENGTH_MASK;
-import static com.oath.oak.Chunk.POSITION_ARRAY_INDEX;
-
-public class OakRKeyBufferImpl implements OakRBuffer {
-
-    private final long keyReference;
+/*
+ * The OakRReference allows reuse of the same OakRBuffer implementation object and is used for
+ * Oak's StreamIterators, where the iterated OakRBuffers can be used only once.
+ * This class is actually a reference into internal BB object rather than new BB object.
+ * It references the internal BB object as far as OakRReference wasn't moved to point on other BB.
+ *
+ * The OakRReference is intended to be used in threads that are for iterations only and are not involved in
+ * concurrent/parallel reading/updating the mappings
+ *
+ * Unlike other ephemeral objects, even if OakRReference references a value it does not have to acquire a read lock
+ * before each access since it can only be used without other concurrent writes in the background.
+ * */
+
+public class OakRKeyBufferImpl implements OakRBuffer, OakUnsafeDirectBuffer {
+
+    private int blockID = OakNativeMemoryAllocator.INVALID_BLOCK_ID;
+    private int position = -1;
+    private int length = -1;
     private final MemoryManager memoryManager;
-    private final int initialPosition;
+    private final int headerSize;
+
+    // The OakRReference user accesses OakRReference as it would be a ByteBuffer with initially zero position.
+    // We translate it to the relevant ByteBuffer position, by adding keyPosition and the header size to any given index
 
-    OakRKeyBufferImpl(long keyReference, MemoryManager memoryManager) {
-        this.keyReference = keyReference;
+    OakRKeyBufferImpl(MemoryManager memoryManager, int headerSize) {
         this.memoryManager = memoryManager;
-        this.initialPosition = UnsafeUtils.longToInts(keyReference)[POSITION_ARRAY_INDEX];
+        this.headerSize = headerSize;
     }
 
-    private ByteBuffer getKeyBuffer() {
-        int[] keyArray = UnsafeUtils.longToInts(keyReference);
```

Comment:
I see, since we are creating an array of ints (object) anyway, no point in storing the data in a single long

Reply:
This is solved by PR https://github.com/yahoo/Oak/pull/80